homeassistant/components/ihc/__init__.py

```diff
@@ -9,38 +9,106 @@
 import xml.etree.ElementTree
 
 import voluptuous as vol
-
+from homeassistant.components.binary_sensor import (
+    DEVICE_CLASSES_SCHEMA)
 from homeassistant.components.ihc.const import (
     ATTR_IHC_ID, ATTR_VALUE, CONF_AUTOSETUP, CONF_BINARY_SENSOR, CONF_DIMMABLE,
-    CONF_INFO, CONF_INVERTING, CONF_LIGHT, CONF_NODE, CONF_SENSOR, CONF_SWITCH,
-    CONF_XPATH, SERVICE_SET_RUNTIME_VALUE_BOOL,
+    CONF_INFO, CONF_INVERTING, CONF_LIGHT, CONF_NODE, CONF_NOTE, CONF_POSITION,
+    CONF_SENSOR, CONF_SWITCH, CONF_XPATH, SERVICE_SET_RUNTIME_VALUE_BOOL,
     SERVICE_SET_RUNTIME_VALUE_FLOAT, SERVICE_SET_RUNTIME_VALUE_INT)
 from homeassistant.config import load_yaml_config_file
 from homeassistant.const import (
-    CONF_ID, CONF_NAME, CONF_PASSWORD, CONF_TYPE, CONF_UNIT_OF_MEASUREMENT,
-    CONF_URL, CONF_USERNAME, TEMP_CELSIUS)
+    CONF_BINARY_SENSORS, CONF_ID, CONF_LIGHTS, CONF_NAME, CONF_PASSWORD,
+    CONF_SENSORS, CONF_SWITCHES, CONF_TYPE, CONF_UNIT_OF_MEASUREMENT, CONF_URL,
+    CONF_USERNAME, TEMP_CELSIUS)
 from homeassistant.helpers import discovery
 import homeassistant.helpers.config_validation as cv
 from homeassistant.helpers.typing import HomeAssistantType
 
 REQUIREMENTS = ['ihcsdk==2.2.0']
 
 DOMAIN = 'ihc'
-IHC_DATA = 'ihc'
+IHC_DATA = 'ihc{}'
 IHC_CONTROLLER = 'controller'
 IHC_INFO = 'info'
 AUTO_SETUP_YAML = 'ihc_auto_setup.yaml'
 
+
+def validate_name(config):
+    """Validate device name."""
+    if CONF_NAME in config:
+        return config
+    ihcid = config[CONF_ID]
+    name = 'ihc_{}'.format(ihcid)
+    config[CONF_NAME] = name
+    return config
+
+
+DEVICE_SCHEMA = vol.Schema({
+    vol.Required(CONF_ID): cv.positive_int,
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Optional(CONF_POSITION): cv.string,
+    vol.Optional(CONF_NOTE): cv.string,
+})
+
+SWITCH_SCHEMA = vol.Schema({
+    vol.All(cv.ensure_list, [
+        [DEVICE_SCHEMA],
+        vol.All({
+            vol.Required(CONF_ID): cv.positive_int,
+            vol.Optional(CONF_NAME): cv.string,
+        }, validate_name)
+    ])
+})
+
+BINARY_SENSOR_SCHEMA = vol.Schema({
+    vol.All(cv.ensure_list, [
```

Comment:
The schema doesn't look right. Looks like it's too many brackets. We want to validate a list of dictionaries.

Reply:
I might have placed it wrong. The intend was to move all the standard settings to DEVICE_SCHEMA and include it in the device schemas. Ie as an alternative to this:
```yaml
BINARY_SENSOR_SCHEMA = vol.Schema({
    vol.All(cv.ensure_list, [
        vol.All({
            vol.Required(CONF_ID): cv.positive_int,
            vol.Optional(CONF_NAME): cv.string,
            vol.Optional(CONF_POSITION): cv.string,
            vol.Optional(CONF_NOTE): cv.string,
            vol.Optional(CONF_TYPE): DEVICE_CLASSES_SCHEMA,
            vol.Optional(CONF_INVERTING, default=False): cv.boolean,
        }, validate_name)
    ])
})
```
What would be the correct way to do the split/include of DEVICE_SCHEMA ?

Reply:
We can use `.extend(...)` on dictionary based schemas to update them with more items.

Reply:
Thanks. I have updated. Will run some configuration tests tomorrow

Reply:
Help needed! I have splittet the schemas and and is extending them for each sensor. So far so good. 
```python
DEVICE_SCHEMA = vol.Schema({
    vol.Required(CONF_ID): cv.positive_int,
    vol.Optional(CONF_NAME): cv.string,
    vol.Optional(CONF_POSITION): cv.string,
    vol.Optional(CONF_NOTE): cv.string
}, extra=vol.ALLOW_EXTRA)

BINARY_SENSOR_SCHEMA = DEVICE_SCHEMA.extend({
    vol.Optional(CONF_TYPE): DEVICE_CLASSES_SCHEMA,
    vol.Optional(CONF_INVERTING, default=False): cv.boolean,
})
```
My trouble starts when adding them to the main configuration schema.
```python
IHC_SCHEMA = vol.Schema({
    vol.Required(CONF_URL): cv.string,
   ...
    vol.Optional(CONF_BINARY_SENSORS, default=[]):
        vol.All(cv.ensure_list, [BINARY_SENSOR_SCHEMA]),
or
        vol.All(cv.ensure_list, [
            vol.All({
                vol.All([BINARY_SENSOR_SCHEMA])
            }, validate_name)
        ]),
```
The code works but i've have lost the validation in each cases :-(
The original code was
```
PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
    vol.Optional(CONF_BINARY_SENSORS, default=[]):
        vol.All(cv.ensure_list, [
            vol.All({
                vol.Required(CONF_ID): cv.positive_int,
                vol.Optional(CONF_NAME): cv.string,
               ...
            }, validate_name)
        ])
})
```
Suggestions?